Error message when knot multiplicity too high for periodic splines

For periodic splines, the error message when increasing the multiplicity over the periodic edge above the spline degree is unclear.

Message tail
----------------
...
 line 1381, in _localsplinebasis
    assert lknots[p]-lknots[p-1]>numpy.spacing(1), 'Element size should be positive'
AssertionError: Element size should be positive

Script to reproduce
------------------------

from nutils import *

periodic_edge_multiplicity = 4  # Should be max "p", as otherwise the spline is open 

domain, geom0 = mesh.rectilinear([8],periodic=[0])
bsplinebasis = domain.basis('spline', degree=(3,), knotmultiplicities=[[periodic_edge_multiplicity, 1, 1, 1, 1, 1, 1, 1, periodic_edge_multiplicity]], knotvalues=[[0.0, 1.5707963267948966, 3.1415926535897931, 4.7123889803846897, 6.2831853071795862, 7.8539816339744828, 9.4247779607693793, 10.995574287564276, 12.566370614359172]])

print('Nr. of. basis functions: {}'.format(bsplinebasis.shape[0]) )

bezier = domain.sample('bezier', 10)
x, N = bezier.eval([geom0, bsplinebasis])
with export.mplfigure('basisfunctions.png') as fig:
  ax = fig.add_subplot(111)
  ax.plot(x, N)
  ax.grid()

--- a/nutils/topology.py
+++ b/nutils/topology.py
@@ -1148,7 +1148,6 @@ class StructuredTopology(Topology):
     for idim in range(self.ndims):
       p = degree[idim]
       n = self.shape[idim]
-      isperiodic = idim in periodic
 
       c = continuity[idim]
       if c < 0:
@@ -1180,7 +1179,7 @@ class StructuredTopology(Topology):
           m = m_
         assert len(m) == n+1, 'knot multiplicity do not match the topology size'
 
-      if isperiodic:
+      if idim in periodic and not m[0] == m[n] == p+1: # if m[0] == m[n] == p+1 the spline is discontinuous at the boundary
         assert m[0] == m[n], 'periodic spline multiplicity expected'
         dk = k[n] - k[0]
         m = m[:n]

--- a/tests/test_basis.py
+++ b/tests/test_basis.py
@@ -221,7 +221,7 @@ for btype in ['discont', 'spline', 'std']:
         for continuity in [-1] if btype == 'discont' else [0] if btype == 'std' else range(degree):
           structured_rect1d(periodic=periodic, btype=btype, degree=degree, nelems=nelems, continuity=continuity)
 
-class structured_rect1d_knotmultiplicities(basisTest):
+class structured_rect1d_periodic_knotmultiplicities(basisTest):
 
   def test(self):
     for knotmultiplicities, ndofs in [([3,1,3], 4), ([3,2,1,3], 6)]:
@@ -230,6 +230,15 @@ class structured_rect1d_knotmultiplicities(basisTest):
       self.assertEqual(len(basis), ndofs)
       self.assertContinuous(topo=domain, geom=geom, basis=basis, continuity=0)
       self.assertPartitionOfUnity(topo=domain, basis=basis)
+
+  def test_discontinuous(self):
+    pdomain, pgeom = mesh.rectilinear([3], periodic=[0])
+    rdomain, rgeom = mesh.rectilinear([3])
+    pbasis = pdomain.basis('spline', degree=2, knotmultiplicities=[[3,1,2,3]])
+    rbasis = rdomain.basis('spline', degree=2, knotmultiplicities=[[1,1,2,1]])
+    psampled = pdomain.sample('gauss', 2).eval(pbasis)
+    rsampled = rdomain.sample('gauss', 2).eval(rbasis)
+    self.assertAllAlmostEqual(psampled, rsampled)
 
 @parametrize
 class unstructured_topology(TestCase):

